f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py

```diff
@@ -233,9 +305,118 @@ def _pre_create(self, service):
+                # from vs
+                vs = self.vs_helper.load(bigip, name=vs_payload['name'],
+                                         partition=vs_payload['partition'])
+                vs_payload['rules'] = vs.rules
+                if add is True:
+                    if irule_name not in vs_payload['rules']:
+                        vs_payload['rules'].append(irule_name)
+                else:
+                    if irule_name in vs_payload['rules']:
+                        vs_payload['rules'].remove(irule_name)
+                self.vs_helper.update(bigip, vs_payload)
+        return
+
+    def __add_bwc(self, bigip, loadbalancer, service, update_listener=True):
+        # The logic is , add the irule bwc policies and go through listeners to
+        # bind the bwc policy.
+        bandwidth = loadbalancer.get('bandwidth')
+        payload = self._create_bwc_payload(loadbalancer, bandwidth)
+        super(LoadBalancerManager, self)._create(
+            bigip, payload, None, None, type="bwc-profile",
+            helper=self.bwc_profile_helper, overwrite=True)
+
+        payload = self._create_irule_payload(loadbalancer)
+        super(LoadBalancerManager, self)._create(
+            bigip, payload, None, None, type="irule-profile",
+            helper=self.irule_helper, overwrite=True)
+
+        # go through the listeners and update them
+        if update_listener is True:
+            self.__update_listener_bwc(bigip, loadbalancer, service)
+        return
+
+    def __delete_bwc(self, bigip, loadbalancer, service):
+        # The logic is , unbond the listenres' bwc policy and delete the irule
+        # bwc policies and go through listeners to
+
+        # go through the listeners and delete them
+        self.__update_listener_bwc(bigip, loadbalancer, service, False)
+        payload = self._create_irule_payload(loadbalancer, True)
+        super(LoadBalancerManager, self)._delete(bigip, payload,
+                                                 None, None,
+                                                 type="irule-profile",
+                                                 helper=self.irule_helper)
+
+        payload = self._create_bwc_payload(loadbalancer, 0)
+        super(LoadBalancerManager, self).\
+            _delete(bigip, payload, None, None,
+                    type="bwc-profile", helper=self.bwc_profile_helper)
+        return
+
+    def __modify_bwc(self, bigip, loadbalancer):
+        # The logic is quit simple with updating the bandwidth
+        bandwidth = loadbalancer.get('bandwidth')
+        payload = self._create_bwc_payload(loadbalancer, bandwidth)
+        super(LoadBalancerManager, self).\
+            _create(bigip, payload, None, None,
+                    type="bwc-profile", helper=self.bwc_profile_helper,
+                    overwrite=True)
+
+    def _update_bwc(self, old_loadbalancer, loadbalancer, service):
+        operate = self.__get_update_operation(old_loadbalancer, loadbalancer)
```

Comment:
__get_update_operation() looks complex, but I feel the logic is not really complex. Do you think the following code looks correct?

```
if old_loadbalancer.get("bandwidth", 0) == loadbalancer.get("bandwidth", 0):
    # Do nothing
else:
    loadbalancer.get("bandwidth", 0) > 0:
        # add_bwc(), which create/update bwc policy and bwc irule, ensure bwc irule is attached to every VS
    loadbalancer.get("bandwidth", 0) <= 0:
        # delete_bwc(), which detach bwc irule from every VS, delete bwc irule and bwc policy (optional)
```